scientific-computing-with-python/learn-tree-traversal-by-building-a-binary-search-tree/binary-search-tree.py

```diff
@@ -0,0 +1,83 @@
+class TreeNode:
+    def __init__(self, key):
+        self.key = key
+        self.left = None
+        self.right = None
+
+class BinarySearchTree:
+    def __init__(self):
+        self.root = None
+
+    def insert(self, key):
+        self.root = self._insert(self.root, key)
```

Comment:
If I'm not mistaken, structuring inserts and deletions like this will actually re-assign multiple nodes, even if that's not needed. The bigger tree, the more re-assigns will occur. That's a bit pointless.

Reply:
For `insert` I was thinking about something like this:
```python
    def insert(self, key):
        new_node = TreeNode(key)
        if self.root is None:
            self.root = new_node
            return
        self._insert(self.root, new_node)

    def _insert(self, parent_node, new_node):
        if new_node.key < parent_node.key:
            if parent_node.left is None:
                parent_node.left = new_node
                return
            self._insert(parent_node.left, new_node)
        elif new_node.key > parent_node.key:
            if parent_node.right is None:
                parent_node.right = new_node
                return
            self._insert(parent_node.right, new_node)
```

Version without recursion:
```python
    def insert(self, key):
        new_node = TreeNode(key)
        if self.root is None:
            self.root = new_node
            return

        parent_node = self.root
        while True:
            if new_node.key < parent_node.key:
                if parent_node.left is None:
                    parent_node.left = new_node
                    return
                parent_node = parent_node.left
            elif new_node.key > parent_node.key:
                if parent_node.right is None:
                    parent_node.right = new_node
                    return
                parent_node = parent_node.right
            else:
                break
```